src/core/namespace.c

```diff
@@ -287,10 +288,15 @@ static int apply_mount(
 
                 /* First, get rid of everything that is below if there
                  * is anything... Then, overmount it with an
-                 * inaccessible directory. */
+                 * inaccessible path. */
                 umount_recursive(m->path, 0);
 
-                what = "/run/systemd/inaccessible";
+                r = lstat(m->path, &inaccessible_target);
```

Comment:
mount() always follow symlinks: a symlink itself cannot be bind-mounted. So the stat should follow the symlink too.

Unfortunately, you cannot rely on `stat()` to follow the symlinks correctly because `apply_mount()` is called before the chroot from [RootDirectory=](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#RootDirectory=).

Slightly related issue about following symlinks https://github.com/systemd/systemd/issues/2860